### PR TITLE
Augment `--allow-{newer,older}` syntax to support wildcards

### PR DESCRIPTION
**Pull request by hvr**
This continues the work started in  a0d8035 (#4575)

Please include the following checklist in your PR:

* [x] Patches conform to the [coding conventions](https://github.com/haskell/cabal/#conventions).
* [x] Any changes that could be relevant to users have been recorded in the changelog.
* [x] The documentation has been updated, if necessary.

Please also shortly describe how you tested your change. Bonus points for added tests!

### CODE REVIEW

File: cabal-install/Distribution/Client/Types.hs [hunk 1]
```diff
@@ -454,9 +466,35 @@ instance Text RelaxedDep where
...
+instance Text RelaxDeps where
+  disp rd | not (isRelaxDeps rd) = Disp.text "none"
+  disp (RelaxDepsSome pkgs)      = Disp.fsep .
+                                   Disp.punctuate Disp.comma .
+                                   map disp $ pkgs
+  disp RelaxDepsAll              = Disp.text "all"
+
+  parse =           (const mempty        <$> ((Parse.string "none" Parse.+++
+                                               Parse.string "None") <* Parse.eof))
+          Parse.<++ (const RelaxDepsAll  <$> ((Parse.string "all"  Parse.+++
```
Comment (23Skidoo) [inline cabal-install/Distribution/Client/Types.hs]: Shouldn't `*` be accepted here as well?

File: cabal-install/tests/UnitTests/Distribution/Client/ProjectConfig.hs [hunk 2]
```diff
@@ -248,6 +249,12 @@ prop_roundtrip_printparse_RelaxedDep rdep =
       [rdep'] -> rdep' == rdep
       _       -> False
 
+prop_roundtrip_printparse_RelaxDeps :: RelaxDeps -> Bool
+prop_roundtrip_printparse_RelaxDeps rdep =
+    case [ x | (x,"") <- Parse.readP_to_S Text.parse (Text.display rdep) ] of
```
Comment (23Skidoo) [inline cabal-install/tests/UnitTests/Distribution/Client/ProjectConfig.hs]: This never generates the `*`-variants, correct? May be worth introducing a helper function that does that.

File: cabal-install/Distribution/Client/Types.hs (hunk 1 above)
Comment (hvr) [inline cabal-install/Distribution/Client/Types.hs]: We can, but it's not essential, as there's two ways to represent `*` currently. In a future refactoring, I do intend to make `--allow-newer` more like `--preferences` and `--constraints`, at which point the outer `RelaxDep` structure will become merely a newtype, which will avoid these corner-cases which make round-tripping tests so weird.

**Comment by ezyang** [comment]
Tests for the new user-visible functionality?!

**Comment by hvr** [comment]
@ezyang indeed, I need to come up with some high-level tests, besides the parser-centric ones.
